# Descripción de los datos
# 
Leer los archivos de datos (/datasets/instacart_orders.csv, /datasets/products.csv, /datasets/aisles.csv, /datasets/departments.csv y /datasets/order_products.csv) con pd.read_csv() usando los parámetros adecuados para leer los datos correctamente.

In [76]:
# importar librerías
import pandas as pd
import os

In [40]:
# leer conjuntos de datos en los DataFrames

orders_df = pd.read_csv('../datasets/raw/instacart_orders.csv', sep=';')
products_df = pd.read_csv('../datasets/raw/products.csv', sep=';')
order_product_df= pd.read_csv('../datasets/raw/order_products.csv', sep=';')
aisles_df= pd.read_csv('../datasets/raw/aisles.csv', sep=';')
departaments_df = pd.read_csv('../datasets/raw/departments.csv', sep=';')

### Descripción de los datos

In [41]:
# mostrar información del DataFrame
print('Tabla de Pedidos')
orders_df.info()
print()
print(orders_df.head())
print()
print('Tabla de Productos')
products_df.info()
print()
print(products_df.head())
print()
print('Tabla de Productos por Pedido')
order_product_df.info()
print()
print(order_product_df.head())
print()
print('Tabla de Pasillos')
aisles_df.info()
print()
print(aisles_df.head())
print()
print('Tabla de Departamentos')
departaments_df.info()
print()
print(departaments_df.head())

Tabla de Pedidos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478967 entries, 0 to 478966
Data columns (total 6 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   order_id                478967 non-null  int64  
 1   user_id                 478967 non-null  int64  
 2   order_number            478967 non-null  int64  
 3   order_dow               478967 non-null  int64  
 4   order_hour_of_day       478967 non-null  int64  
 5   days_since_prior_order  450148 non-null  float64
dtypes: float64(1), int64(5)
memory usage: 21.9 MB

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   1515936   183418            11          6                 13   
1   1690866   163593             5          5                 12   
2   1454967    39980             4          5                 19   
3   1768857    82516            56          0                 20   
4   3007858   196724             2          4    

### Duplicados y valores ausentes

In [42]:
print('Numero de duplicados:',orders_df.duplicated().sum())
print()
duplicated_orders_mask = orders_df.duplicated()
print(orders_df[duplicated_orders_mask])





Numero de duplicados: 15

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
145574    794638    50898            24          3                  2   
223105   2160484   107525            16          3                  2   
230807   1918001   188546            14          3                  2   
266232   1782114   106752             1          3                  2   
273805   1112182   202304            84          3                  2   
284038   2845099    31189            11          3                  2   
311713   1021560    53767             3          3                  2   
321100    408114    68324             4          3                  2   
323900   1919531   191501            32          3                  2   
345917   2232988    82565             1          3                  2   
371905    391768    57671            19          3                  2   
394347    467134    63189            21          3                  2   
411408   1286742   183220

In [43]:
print('Numero de duplicados:',products_df.duplicated().sum())
print()
print('Duplicados implicitos:',products_df['product_name'].duplicated().sum())
print()
duplicated_products_mask = products_df['product_name'].duplicated()
print(products_df[duplicated_products_mask])



Numero de duplicados: 0

Duplicados implicitos: 1257

       product_id product_name  aisle_id  department_id
71             72          NaN       100             21
109           110          NaN       100             21
296           297          NaN       100             21
416           417          NaN       100             21
436           437          NaN       100             21
...           ...          ...       ...            ...
49552       49553          NaN       100             21
49574       49575          NaN       100             21
49640       49641          NaN       100             21
49663       49664          NaN       100             21
49668       49669          NaN       100             21

[1257 rows x 4 columns]


In [44]:
print(order_product_df.isna().sum())
print()
print('Numero de duplicados:',order_product_df.duplicated().sum())



order_id               0
product_id             0
add_to_cart_order    836
reordered              0
dtype: int64

Numero de duplicados: 0


In [45]:
print()
print('Numero de duplicados:',aisles_df.duplicated().sum())
print()
print('Duplicados implicitos:',aisles_df['aisle'].duplicated().sum())




Numero de duplicados: 0

Duplicados implicitos: 0


In [46]:
print()
print('Numero de duplicados:',departaments_df.duplicated().sum())
print('Duplicados implicitos:', departaments_df['department'].duplicated().sum())


Numero de duplicados: 0
Duplicados implicitos: 0


1.  orders_df   
En este DataFrame tenemos las siguintes observaciones:

Los nombres de las columnas tienen un buen formato, por lo que no es necesario realizarles algun cambio,
Este DataFrame solamente tiene una columna con datos ausentes por lo que sera necesario revisarlo a detalle más adelante
La columna 'days_since_prior_order' tiene datos de tipo flotante, dado que es un conteo de días, considero que no deberia ser de tipo flotante sino de tipo entero, ya que no hay dias fraccionados.
En el DataFrame existen 15 filas duplicadas, por lo que sugiere algun tipo de error en este día y hora en especifico.

2.  products_df   
En este DataFrame tenemos las siguintes observaciones:

Los nombres de las columnas tienen un buen formato, por lo que no es necesario realizarles algun cambio.
La columna 'product_name' cuenta con datos asusentes o nulos, por lo que sera necesario revisarlo más adelante.
No se encuentran duplicados explicitos, sin embargo al revisar 'product_name' observamos que arrojan duplicados pero por valores ausentes en la misma columna.

3.  order_product_df   
En este DataFrame tenemos las siguintes observaciones:

Los nombres de las columnas tienen un buen formato, por lo que no es necesario realizarles algun cambio.
La columna 'add_to_cart_order' tiene datos de tipo flotante, dado que es un orden de productos, considero que no deberia ser de tipo flotante sino de tipo entero, ya que no hay conteo fraccionado.
el metodo info() no me muestra el conteo de entradas no nulas, sin embargo podemos realizar el conteo con el metodo isna(), ahora ya podemos observar que hay 836 valores nulos en la columna 'add_to_cart_order'

4.  aisles_df   
En este DataFrame tenemos las siguintes observaciones:

Podemos observar que tanto los nombres de las columnas y los tipos de datos son consistentes, y no existen valores nulos en esta DataFrame

5.  departaments_df   
En este DataFrame tenemos las siguintes observaciones:

Podemos observar que tanto los nombres de las columnas y los tipos de datos son consistentes, y no existen valores nulos en esta DataFrame
Como conclusion general, todas los DataFrame creados tienen nombres de columnas consistentes, algunos datos podriamos cambiarlos a otro tipo de dato para que sea congruente con la información, y existen valores nulos que debemos revisar para determinar que realizar con ellos, ademas de que encontramos duplicados explicitos e implicitos que más adelante corregiremos.

# Preprocesamiento de los datos

Antes de realizar nuestro análisis lo que haremos sera preparar los datos corrigiendo los tipos de datos, valores ausentes y valores duplicados,

Primero corregiremos las columnas, 'days_since_prior_order'y 'add_to_cart_order' con un tipo de dato más adecuado ya que estan en tipo flotante y lo cambiaremos a tipo entero con el metodo astype()

Despues identificaremos duplicados explícitos dentro de los DataFrame y asi poder eliminarlos de las tablas con los metodos duplicated() y drop_duplicates(), luego procederemos a buscar duplicados implicitos que puedan existir dentro de las columnas con el metodo unique() y verificar que no existan datos erroneos como algun duplicado por una mayuscula o minuscula que pueda afectar a los datos.

Finalmente procesaremos los valores ausentes de las columnas 'product_name', 'days_since_prior_order' y 'add_to_cart_order' de la tablas products, orders y order_productos respectivamente.

## Valores duplicados

### Dataframe Orders

In [47]:
# Revisamos si hay pedidos duplicados
print('Numero de filas duplicadas', orders_df.duplicated().sum())
print()
print(orders_df[orders_df.duplicated()])

Numero de filas duplicadas 15

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
145574    794638    50898            24          3                  2   
223105   2160484   107525            16          3                  2   
230807   1918001   188546            14          3                  2   
266232   1782114   106752             1          3                  2   
273805   1112182   202304            84          3                  2   
284038   2845099    31189            11          3                  2   
311713   1021560    53767             3          3                  2   
321100    408114    68324             4          3                  2   
323900   1919531   191501            32          3                  2   
345917   2232988    82565             1          3                  2   
371905    391768    57671            19          3                  2   
394347    467134    63189            21          3                  2   
411408   1286742   1

Observamos que todos los pedidos se hicieron el miercoles a las 2 de la mañana.

In [48]:
# Verificamos que todos los pedidos que se hicieron el miércoles a las 2:00 a.m.

print(orders_df.query("order_dow == 3 & order_hour_of_day == 2"))

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
4838     2766110   162084            41          3                  2   
5156     2190225   138285            18          3                  2   
15506     553049    58599            13          3                  2   
18420     382357   120200            19          3                  2   
24691     690242    77357             2          3                  2   
...          ...      ...           ...        ...                ...   
457013   3384021    14881             6          3                  2   
458816    910166   164782            18          3                  2   
459635   1680532   106435             6          3                  2   
468324    222962    54979            59          3                  2   
477526   2592344    46860            38          3                  2   

        days_since_prior_order  
4838                      16.0  
5156                      11.0  
15506                   

Lo que observamos es que debio existir algun tipo de error por parte de la aplicación, ya sea por comunicación fallida que duplico las entradas en este día de la semana, ya que no existen duplicados más alla de otros días.

In [49]:
# se elimina los pedidos duplicados
orders_df = orders_df.drop_duplicates().reset_index(drop=True)

In [50]:
# verificamos si hay filas duplicadas
print(orders_df.duplicated().sum())

0


In [51]:
# Verificamos únicamente si hay IDs duplicados de pedidos
print(orders_df['order_id'].duplicated().sum())

0


Descubrimos que existio un tipo de error especificamente el día miercoles a las 2 am, que permitio crear duplicados en nuestra tabla.

### Dataframe products

In [52]:
# Verificamos si hay filas totalmente duplicadas
print('Numero de filas duplicadas', products_df.duplicated().sum())

Numero de filas duplicadas 0


In [53]:
# Revisamos si hay ID de departamentos duplicados
print(products_df['department_id'].duplicated().sum())

49673


In [54]:
# Revisamos si hay nombres duplicados de productos

products_df['product_name'] = products_df['product_name'].str.upper()
print(products_df)
print(products_df['product_name'].duplicated().sum())

       product_id                                       product_name  \
0               1                         CHOCOLATE SANDWICH COOKIES   
1               2                                   ALL-SEASONS SALT   
2               3               ROBUST GOLDEN UNSWEETENED OOLONG TEA   
3               4  SMART ONES CLASSIC FAVORITES MINI RIGATONI WIT...   
4               5                          GREEN CHILE ANYTIME SAUCE   
...           ...                                                ...   
49689       49690                      HIGH PERFORMANCE ENERGY DRINK   
49690       49691                      ORIGINAL PANCAKE & WAFFLE MIX   
49691       49692    ORGANIC INSTANT OATMEAL LIGHT MAPLE BROWN SUGAR   
49692       49693                             SPRING WATER BODY WASH   
49693       49694                            BURRITO- STEAK & CHEESE   

       aisle_id  department_id  
0            61             19  
1           104             13  
2            94              7  
3  

In [55]:
# Revisamos si hay nombres duplicados de productos no faltantes

filter_product = products_df[~products_df['product_name'].isna()]
print(filter_product['product_name'].duplicated().sum())
print()
products_df = products_df.drop_duplicates(subset='product_name').reset_index(drop=True)
print(products_df['product_name'].duplicated().sum())

104

0


En la columna product_name sin procesar existen 1257 entradas duplicadas y al realizar el cambio a mayusculas aumento el numero de dulicados esto porque en algunos casos no encontro duplicados implicitos, al colocar todo en mayusculas se puede obeervar mejor los duplicados existentes que son 1361, ademas vemos que 104 valores vienen de los productos no faltantes sin embargo la mayoria de estos duplicados son de valores ausentes, al filtrar quitando los valores no faltantes, eliminamos los duplicados por la columna y volvemos a revisar que quede en cero.

### Dataframe departments

In [56]:
# Revisamos si hay filas totalmente duplicadas
print('Numero de filas duplicadas', departaments_df.duplicated().sum())

Numero de filas duplicadas 0


In [57]:
# Revisamos únicamente si hay IDs duplicadas de productos
print('Numero de filas duplicadas', departaments_df['department_id'].duplicated().sum())

Numero de filas duplicadas 0


En la tabla 'departaments' no se arrojan duplicados.

### Dataframe aisles

In [58]:
# Revisamos si hay filas totalmente duplicadas
print('Numero de filas duplicadas', aisles_df.duplicated().sum())

Numero de filas duplicadas 0


In [59]:
# Revisamos únicamente si hay IDs duplicadas de productos
print('Numero de filas duplicadas', aisles_df['aisle_id'].duplicated().sum())

Numero de filas duplicadas 0


En la tabla 'aisles' no se encontraron duplicados.

### Dataframe order_products

In [60]:
# Revisamos si hay filas totalmente duplicadas
print('Numero de filas duplicadas', order_product_df.duplicated().sum())

Numero de filas duplicadas 0


In [61]:
# Volvemos a verificar si hay cualquier otro duplicado engañoso
print('Numero de filas duplicadas', order_product_df['order_id'].duplicated().sum())
print('Numero de filas duplicadas', order_product_df['product_id'].duplicated().sum())
print('Numero de filas duplicadas', order_product_df['add_to_cart_order'].duplicated().sum())
print('Numero de filas duplicadas', order_product_df['reordered'].duplicated().sum())

Numero de filas duplicadas 4094961
Numero de filas duplicadas 4499434
Numero de filas duplicadas 4544942
Numero de filas duplicadas 4545005


En la tabla 'order_products' no existen duplicados, sin embargo al buscar duplicados por columna nos arrojan resultados, pero estos se duplican por el hecho de que cada orden puede tener varios productos.

## Valores ausentes

Al trabajar con valores duplicados, pudimos observar que también nos falta investigar valores ausentes:

1. La columna 'product_name' de la tabla products.
2. La columna 'days_since_prior_order' de la tabla orders.
3. La columna 'add_to_cart_order' de la tabla order_productos.

### Dataframe products

In [62]:
# Buscaremos los valores ausentes en la columna 'product_name'
print(products_df.isna().sum())
print()
print(products_df[products_df['product_name'].isna()])

product_id       0
product_name     1
aisle_id         0
department_id    0
dtype: int64

    product_id product_name  aisle_id  department_id
37          38          NaN       100             21


En la primera parte al revisar duplicados observamos que habian 104 valores duplicados ausentes, se eliminaron por lo que ahora solo queda un valor ausente que no estaba duplicado.

In [63]:
#  ¿Todos los nombres de productos ausentes están relacionados con el pasillo con ID 100?
print(products_df[products_df['aisle_id'] == 100])

    product_id product_name  aisle_id  department_id
37          38          NaN       100             21


Nos arrojan los mismos resultados al filtrar por pasillo #100.

In [64]:
# ¿Todos los nombres de productos ausentes están relacionados con el departamento con ID 21?
print(products_df[products_df['department_id'] == 21])

    product_id product_name  aisle_id  department_id
37          38          NaN       100             21


Nos vuelve a arrojar el mismo resultado al filtrar por el departamento #21.

In [65]:
print(products_df[(products_df['aisle_id'] == 100 ) & (products_df['department_id'] == 21)])

    product_id product_name  aisle_id  department_id
37          38          NaN       100             21


Los valores ausentes de la columna 'product_name' se encuentran relacionados con el pasillo numero 100 y departamento 21.

In [66]:
# Completamos los nombres de productos ausentes con 'Unknown'
products_df['product_name'] = products_df['product_name'].fillna('Unknown')
print(products_df.isna().sum())

product_id       0
product_name     0
aisle_id         0
department_id    0
dtype: int64


Descubrimos que los valores ausentes se relacionaban estrechamente con el pasillo 100 y departmento 21, por lo que hemos remplazados los valores ausentes de la columna 'product_name' por 'Unknown' y al volver a contar observamos que se realizo el remplazo correctamente.

### Dataframe orders

In [67]:
# buscamos los valores ausentes
print(orders_df.isna().sum())
print()
print(orders_df[orders_df['days_since_prior_order'].isna()])

order_id                      0
user_id                       0
order_number                  0
order_dow                     0
order_hour_of_day             0
days_since_prior_order    28817
dtype: int64

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
28        133707   182261             1          3                 10   
96        787445    25685             1          6                 18   
100       294410   111449             1          0                 19   
103      2869915   123958             1          4                 16   
104      2521921    42286             1          3                 18   
...          ...      ...           ...        ...                ...   
478880   2589657   205028             1          0                 16   
478881   2222353   141211             1          2                 13   
478907   2272807   204154             1          1                 15   
478911   2499542    68810             1          4              

In [68]:
# ¿Hay algún valor ausente que no sea el primer pedido del cliente?
print(orders_df[orders_df['order_number'] == 1])
orders_df['days_since_prior_order'] = orders_df['days_since_prior_order'].fillna(0)
print(orders_df.isna().sum())

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
28        133707   182261             1          3                 10   
96        787445    25685             1          6                 18   
100       294410   111449             1          0                 19   
103      2869915   123958             1          4                 16   
104      2521921    42286             1          3                 18   
...          ...      ...           ...        ...                ...   
478880   2589657   205028             1          0                 16   
478881   2222353   141211             1          2                 13   
478907   2272807   204154             1          1                 15   
478911   2499542    68810             1          4                 19   
478930   1387033    22496             1          5                 14   

        days_since_prior_order  
28                         NaN  
96                         NaN  
100                     

order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64


In [69]:
# Cambiar la columan days_since_prior_order a tipo entero
# order_product_df['add_to_cart_order'] = order_product_df['add_to_cart_order'].astype('int')
orders_df['days_since_prior_order'] = orders_df['days_since_prior_order'].astype('int')

Encontramos que el primer pedido de cada cliente no tiene un numero especificado ya que es justamenete el primer pedido y no existe un conteo antes de este, por lo que podemos remplazarlo com un valor de cero para que no nos afecte y que sea clara la información que estamos interpretando.

### Dataframe order_products

In [70]:
# Buscar los valores ausentes
print(order_product_df.isna().sum())
print()
print(order_product_df[order_product_df['add_to_cart_order'].isna()])

order_id               0
product_id             0
add_to_cart_order    836
reordered              0
dtype: int64

         order_id  product_id  add_to_cart_order  reordered
737       2449164        5068                NaN          0
9926      1968313       43867                NaN          0
14394     2926893       11688                NaN          0
16418     1717990        4142                NaN          0
30114     1959075       42828                NaN          1
...           ...         ...                ...        ...
4505662   1800005        7411                NaN          0
4511400   1633337         260                NaN          0
4517562    404157        9517                NaN          0
4534112   1673227       17835                NaN          0
4535739   1832957       17949                NaN          1

[836 rows x 4 columns]


In [71]:
# ¿Cuáles son los valores mínimos y máximos en esta columna?
print(order_product_df['add_to_cart_order'].min())
print()
print(order_product_df['add_to_cart_order'].max())

1.0

64.0


El conteo empieza en 1 y el maximo valor es el 999.

In [72]:
# Guardamos todas las IDs de pedidos que tengan un valor ausente en 'add_to_cart_order'
nan_order_df = order_product_df[order_product_df['add_to_cart_order'].isna()][['order_id']]
print(nan_order_df)

         order_id
737       2449164
9926      1968313
14394     2926893
16418     1717990
30114     1959075
...           ...
4505662   1800005
4511400   1633337
4517562    404157
4534112   1673227
4535739   1832957

[836 rows x 1 columns]


In [73]:
# ¿Todos los pedidos con valores ausentes tienen más de 64 productos?
nan_order_filtered = order_product_df.query("order_id in @nan_order_df.order_id") 

# Agrupamos todos los pedidos con datos ausentes por su ID de pedido.
nan_order_group = nan_order_filtered.groupby('order_id')['product_id'].count()

# Cuenta el número de 'product_id' en cada pedido y revisa el valor mínimo del conteo.
min_product_count = nan_order_group.min()
max_product_count = nan_order_group.max()
print(nan_order_group)
print(min_product_count)
print(max_product_count)

order_id
9310        65
61355      127
102236      95
129627      69
165801      70
          ... 
2999801     70
3125735     86
3308010    115
3347453     71
3383594     69
Name: product_id, Length: 70, dtype: int64
65
127


Observamos que los pedidos con valores ausentes tienen más de 64 productos. Se tendria que realizar algun análisis mas complejo para determinar la causa de los valores ausentes. Sin embargo es posible que el sistema tenga una limitación al procesar pedidos mayores a 64 productos.

In [74]:
# Remplazaremos los valores ausentes en la columna 'add_to_cart? con 999 y convertiremos la columna al tipo entero.
order_product_df['add_to_cart_order'] = order_product_df['add_to_cart_order'].fillna(999)
order_product_df['add_to_cart_order'] = order_product_df['add_to_cart_order'].astype('int')
print(order_product_df.isna().sum())

order_product_df.info()

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4545007 entries, 0 to 4545006
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 138.7 MB


Descubrimos que hay valores ausentes causados por diferentes problemas en el procesamiento de estos datos, como en el caso del primer pedido que se genera al no tener un valor inicial, tambien en ordenes mayores de 64 productos, por lo que seria importante verificar y soluciar estos errores en la aplicación.

Tambien existe un error en el tipo de datos, por lo que es importante verificar que tipo de datos corresponde a cada columna de las tablas y que su procesamiento sea mas clara y congruente con lo que se esta presentando.

In [75]:
# mostrar información del DataFrame
print('Tabla de Pedidos')
orders_df.info()
print()
print(orders_df.head())
print()
print('Tabla de Productos')
products_df.info()
print()
print(products_df.head())
print()
print('Tabla de Productos por Pedido')
order_product_df.info()
print()
print(order_product_df.head())
print()
print('Tabla de Pasillos')
aisles_df.info()
print()
print(aisles_df.head())
print()
print('Tabla de Departamentos')
departaments_df.info()
print()
print(departaments_df.head())

Tabla de Pedidos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478952 entries, 0 to 478951
Data columns (total 6 columns):
 #   Column                  Non-Null Count   Dtype
---  ------                  --------------   -----
 0   order_id                478952 non-null  int64
 1   user_id                 478952 non-null  int64
 2   order_number            478952 non-null  int64
 3   order_dow               478952 non-null  int64
 4   order_hour_of_day       478952 non-null  int64
 5   days_since_prior_order  478952 non-null  int64
dtypes: int64(6)
memory usage: 21.9 MB

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   1515936   183418            11          6                 13   
1   1690866   163593             5          5                 12   
2   1454967    39980             4          5                 19   
3   1768857    82516            56          0                 20   
4   3007858   196724             2          4                 12   

   days_

In [77]:
# ya que tenemos lo datos limpios los pasamos a la carpeta de datasets para 
# realizar el proceso de EDA
orders_df.to_csv('../datasets/clean/orders_df_clean.csv', index=False)
products_df.to_csv('../datasets/clean/products_df_clean.csv', index=False)
order_product_df.to_csv('../datasets/clean/order_product_df_clean.csv', index=False)
aisles_df.to_csv('../datasets/clean/aisles_df_clean.csv', index=False)
departaments_df.to_csv('../datasets/clean/departaments_df_clean.csv', index=False)
